c/tskit/trees.c

```diff
@@ -3639,6 +3639,191 @@ tsk_tree_prev(tsk_tree_t *self)
+ * use a general cost matrix, in which case we'll use the Sankoff algorithm. For
+ * now this is unused.
+ */
+int TSK_WARN_UNUSED
+tsk_tree_map_mutations(tsk_tree_t *self, int8_t *genotypes,
+        double *TSK_UNUSED(cost_matrix), tsk_flags_t TSK_UNUSED(options),
+        int8_t *r_ancestral_state,
+        tsk_size_t *r_num_transitions, tsk_state_transition_t **r_transitions)
+{
+    int ret = 0;
+    const tsk_size_t num_samples = self->tree_sequence->num_samples;
+    const tsk_size_t num_nodes = self->num_nodes;
+    const tsk_id_t *restrict left_child = self->left_child;
+    const tsk_id_t *restrict right_sib = self->right_sib;
+    const tsk_id_t *restrict parent = self->parent;
+    const tsk_flags_t *restrict node_flags = self->tree_sequence->tables->nodes.flags;
+    uint64_t root_state;
+    int8_t *restrict state = malloc(num_nodes * sizeof(*state));
+    uint64_t *restrict A = calloc(num_nodes, sizeof(*A));
+    tsk_id_t *restrict stack = malloc(num_nodes * sizeof(*stack));
+    tsk_id_t *restrict parent_stack = malloc(num_nodes * sizeof(*parent_stack));
+    tsk_id_t postorder_parent, transition_parent, root, u, v;
+    /* The largest possible number of transitions is one over every sample */
+    tsk_state_transition_t *transitions = malloc(num_samples * sizeof(*transitions));
+    tsk_size_t j;
+    int8_t ancestral_state;
+    int stack_top;
+    tsk_size_t num_transitions;
+    tsk_size_t non_missing = 0;
+
+    if (A == NULL || stack == NULL || parent_stack == NULL || state == NULL
+            || transitions == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    for (j = 0; j < num_samples; j++) {
+        if (genotypes[j] >= 64 || genotypes[j] < TSK_MISSING_DATA) {
+            ret = TSK_ERR_BAD_GENOTYPE;
+            goto out;
+        }
+        u = self->tree_sequence->samples[j];
+        if (genotypes[j] == TSK_MISSING_DATA) {
+            /* All bits set */
+            A[u] = UINT64_MAX;
+        } else {
+            A[u] = set_bit(A[u], genotypes[j]);
+            non_missing++;
+        }
+    }
+
+    if (non_missing == 0) {
+        ret = TSK_ERR_GENOTYPES_ALL_MISSING;
+        goto out;
+    }
+
+    root_state = 0;
+    for (root = self->left_root; root != TSK_NULL; root = self->right_sib[root]) {
+        /* Do a post order traversal */
+        stack[0] = root;
+        stack_top = 0;
+        postorder_parent = TSK_NULL;
+        while (stack_top >= 0) {
+            u = stack[stack_top];
+            if (left_child[u] != TSK_NULL && u != postorder_parent) {
+                for (v = left_child[u]; v != TSK_NULL; v = right_sib[v]) {
+                    stack_top++;
+                    stack[stack_top] = v;
+                }
+            } else {
+                stack_top--;
+                postorder_parent = parent[u];
+
+                /* Visit u */
+                if (! (node_flags[u] & TSK_NODE_IS_SAMPLE)) {
+                    /* Get the union of the child sets */
```

Comment:
you mean "intersection"; and in line n+6 switch 'union' and 'intersection'.

Reply:
Hah, yes, so I did!